package/MDAnalysis/core/universe.py

```diff
@@ -737,11 +737,22 @@ def __repr__(self):
         return "<Universe with {n_atoms} atoms>".format(
             n_atoms=len(self.atoms))
 
-    def __getstate__(self):
-        raise NotImplementedError
+    @classmethod
+    def _unpickle_U(cls, top, traj, anchor):
+        """Special method used by __reduce__ to deserialise a Universe"""
+        # top is a Topology object at this point, but Universe can handle that
+        u = cls(top)
+        u.anchor_name = anchor
+        # maybe this is None, but that's still cool
+        u.trajectory = traj
+
+        return u
 
-    def __setstate__(self, state):
-        raise NotImplementedError
+    def __reduce__(self):
+        # Can't quite use __setstate__/__getstate__ so go via __reduce__
```

Comment:
Why does reduce work and setstate/getstate not? It would be nice to leave a little explanation about what failed to work.

Reply:
I managed to get set/getstate to work.
```python
def __getstate__(self):
    return self._topology, self._trajectory, self.anchor_name
def __setstate__(self, state):
    self._anchor_name = state[2] 
    self._topology = state[0]
    _generate_from_topology(self)
    self._trajectory = state[1]
```

But the result is slightly different.
For analysis of an `AtomGroup`, we need to define it inside the analysis function.
e.g. 
```python
def cog(u, frame_id):
    u.trajectory[frame_id]
    ag = u.atoms[2:5]
    return ag.center_of_geometry()

res = np.array([p.apply(cog, args=(u,i))
            for i in range(3)])
```
gets the same result as before:
```python
def cog(u, ag, frame_id):
    u.trajectory[frame_id]
    return ag.center_of_geometry()

ag = u.atoms[2:5]
res = np.array([p.apply(cog, args=(u,ag, i))
            for i in range(3)])
```

Not sure I am missing something here.

Reply:
can you describe in more detail what the differences are? A quick explanation of the code would also be nice. I do not understand what `p.apply` is.  What happens if you use the pickle functions directly yourself?

Reply:
I am still wrapping  my head around for the difference. The difference seems to be which `Universe` the `AtomGroup` is referencing to after pickling.
```python
def print_id(u, ag, frame_id):
    print(frame_id)
    print('id u:',id(u))
    print('id ag:',id(ag))
    print('id ag._u:',id(ag._u))

u = mda.Universe(PDB_multiframe)
ag = u.atoms[2:5]
print('before pickle')
print('id u:',id(u))
print('id ag:',id(ag))
print('id ag_u:',id(ag._u))

p = multiprocessing.Pool(2)
res = np.array([p.apply(print_id, args=(u, ag, i))
            for i in range(3)])
p.close()
```
- using `__reduce__`:
```

before pickle
id u: 140587994220432
id ag: 140587994220336
id ag_u: 140587994220432
0
id u: 140587993605936
id ag: 140587993697152
id ag._u: 140587993605936
1
id u: 140587993606176
id ag: 140587993697392
id ag._u: 140587993606176
2
id u: 140588514411568
id ag: 140588514555936
id ag._u: 140588514411568
```
- using `__set/getstate__`:
```
before pickle
id u: 140397192364480
id ag: 140397192364384
id ag_u: 140397192364480
0
id u: 140397952037936
id ag: 140397192077952
id ag._u: 140397192364480
1
id u: 140397952037936
id ag: 140397192078192
id ag._u: 140397192364480
2
id u: 140397192523440
id ag: 140397192244960
id ag._u: 140397192364480
```

The `p.apply` is part of the multiprocessing. So in each process, `Universe` and `Atomgroup` are pickled.

Reply:
Your `__setstate__` code is directly setting `_anchor_name`, while `__reduce__` is using the `anchor_name` setter, which additionally runs `remove_anchor()` and `make_anchor()`, so I'm guessing your not seeing ag._u change appropriately has to do with these not being run.

(`remove_anchor` assumes that `self._anchor_name` already exists, though, and I'm not sure it's recreated anywhere else in `__setstate__`, so you might still need to say just set it to None first, or there might be some other workaround)

Reply:
That makes sense! So setting by `anchor_name` get the "right" result.
I also find that because `anchor_name` links between `ag` and `u`, we have to first serialize `u` then `ag`, otherwise, the linkage will break (like above).

Reply:
Please add the comment that explains why `__reduce__`.

Or if you want to rewrite with `__setstate__` then do that :-).